src/topology/homotopy/pi_n.lean

```diff
@@ -0,0 +1,240 @@
+/-
+Copyright (c) 2021 Roberto Alvarez. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Roberto Alvarez
+-/
+
+import topology.homotopy.basic
+import topology.path_connected
+import topology.unit_interval
+import algebraic_topology.fundamental_groupoid.basic
+import logic.unique
+
+/-!
+# nth homotopy group
+
+We define the nth homotopy group at x `π n x` as the equivalence classes
+of functions from the nth dimensional cube to the topological space X
+that send the boundary to the base point `x`, up to homotopic equivalence.
+Note that such functions are generalized loops `gen_loop n x`, in particular
+`gen_loop 1 x ≃ path x x`
+
+We show that `π 0 x` is equivalent to the path-conected components, and
+that `π 1 x` is equivalent to the fundamental grupoid at `x`.
+
+## definitions
+
+* `gen_loop n x` is the type of continous fuctions `I^n→X` that send the boundary to `x`
+* `homotopy_group n x` denoted `π n x` is the quotient of `gen_loop n x` up to homotopic
+  equivalence
+
+-/
+
+open_locale unit_interval topological_space
+
+noncomputable theory
+
+universes u
+variables {X:Type u} [topological_space X]
+variables {n:nat} {x:X}
+
+/--
+The nth dimensional cube
+-/
+@[derive [has_zero, has_one]]
+def cube (n:nat) : Type := fin n → I
+notation `I^` := cube
+
+namespace cube
+instance topology {n} : topological_space (I^n) :=
+  Pi.topological_space
+
+@[continuity] lemma proj_continuous {n} (i:fin n) : continuous (λ (f:I^n), f i) :=
+begin
+  apply continuous_infi_dom,
+  apply continuous_def.2,
+  intros s H,
+  convert ((@is_open_induced_iff' _ _ _ _ (λ (f : I^n), f i)).2 ⟨s, ⟨H, by refl⟩⟩)
+end
+
+/--
+The points of the nth dimensional cube with a projection set to 0 or 1
+-/
+def boundary (n:nat) : set (I^n) :=
+  { y | ∃ i:fin n , y i = 0 ∨ y i = 1}
+
+/--
+The first projection of a non-zero dimensional cube
+-/
+@[simp] def head {n} : I^(n+1) → I := λ c, c 0
+
+@[continuity] lemma head.continuous {n} : continuous (@head n) := proj_continuous 0
+
+/--
+The projection to the last n coordinates from an n+1 dimensional cube
+-/
+@[simp] def tail {n} : I^(n+1) → I^n := λ c, fin.tail c
+
+instance unique_cube0 : unique (I^0) :=
+  { default := 0, uniq := by {intro a, ext1, exact x.elim0} }
+
+lemma one_indep {α} (f : fin 1 → α) (i j : fin 1) : f i = f j :=
+  by rw subsingleton.elim i j
+
+lemma one_char {α} (x : fin 1 → α) : x = λ _, x 0 :=
```

Comment:
(Untested) this is just `unique.eq_default`, since the RHS is def-eq to the default of the type `fin 1 → α`.

Reply:
My bad.